File: backend/apps/analysis/behavioral_model.py
"""
Python port of src/lib/behavioralModel.ts

Provides spending habit analysis, anomaly detection, and forecasting
based on a user's transaction history.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from typing import Any


EXAM_WEEK_KEYWORDS = {"exam", "midterm", "final", "quiz", "project"}


# ---------------------------------------------------------------------------
# Data structures
# ---------------------------------------------------------------------------


@dataclass
class HabitProfile:
    avg_spend_by_day_of_week: dict[int, float] = field(default_factory=dict)
    avg_spend_by_month: dict[int, float] = field(default_factory=dict)
    exam_week_delta: float = 0.0


@dataclass
class CategoryForecast:
    category: str
    amount: float


@dataclass
class HabitualPrediction:
    date: str  # ISO-8601
    predicted_spend: float


@dataclass
class SpendingAnomaly:
    transaction_id: str
    category: str
    amount: float
    z_score: float


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _mean(values: list[float]) -> float:
    if not values:
        return 0.0
    return sum(values) / len(values)


def _std_dev(values: list[float]) -> float:
    if len(values) < 2:
        return 0.0
    avg = _mean(values)
    variance = sum((v - avg) ** 2 for v in values) / len(values)
    return math.sqrt(variance)


def _only_expenses(transactions: list[dict[str, Any]]) -> list[dict[str, Any]]:
    return [t for t in transactions if t.get("type") == "EXPENSE"]


def _normalize_category(category: str | None) -> str:
    return (category or "Other").strip()


def _is_exam_week_transaction(tx: dict[str, Any]) -> bool:
    text = f"{tx.get('description', '')} {tx.get('category', '')}".lower()
    return any(kw in text for kw in EXAM_WEEK_KEYWORDS)


def _get_category_stats(transactions: list[dict[str, Any]]) -> dict[str, dict[str, float]]:
    grouped: dict[str, list[float]] = {}
    for tx in transactions:
        cat = _normalize_category(tx.get("category"))
        grouped.setdefault(cat, []).append(float(tx.get("amount", 0)))

    return {
        cat: {"mean": _mean(vals), "std_dev": _std_dev(vals)}
        for cat, vals in grouped.items()
    }


def _to_datetime(value: Any) -> datetime:
    if isinstance(value, datetime):
        return value
    if isinstance(value, str):
        for fmt in ("%Y-%m-%dT%H:%M:%S.%fZ", "%Y-%m-%dT%H:%M:%SZ", "%Y-%m-%d"):
            try:
                dt = datetime.strptime(value, fmt)
                return dt.replace(tzinfo=timezone.utc)
            except ValueError:
                continue
        # Fallback: isoformat with timezone info
        from dateutil import parser as dateutil_parser  # type: ignore
        try:
            return dateutil_parser.parse(value)
        except Exception:
            return datetime.now(tz=timezone.utc)
    return datetime.now(tz=timezone.utc)


# ---------------------------------------------------------------------------
# Public functions (matching behavioralModel.ts API)
# ---------------------------------------------------------------------------


def learn_habits(transactions: list[dict[str, Any]]) -> HabitProfile:
    """Compute average spending by day-of-week and month, plus exam-week delta."""
    expenses = _only_expenses(transactions)
    day_buckets: dict[int, list[float]] = {i: [] for i in range(7)}
    month_buckets: dict[int, list[float]] = {i: [] for i in range(12)}

    for tx in expenses:
        dt = _to_datetime(tx.get("date"))
        day_buckets[dt.weekday()].append(float(tx.get("amount", 0)))
        month_buckets[dt.month - 1].append(float(tx.get("amount", 0)))

    avg_by_dow = {i: _mean(day_buckets[i]) for i in range(7)}
    avg_by_month = {i: _mean(month_buckets[i]) for i in range(12)}

    exam_expenses = [float(t["amount"]) for t in expenses if _is_exam_week_transaction(t)]
    non_exam_expenses = [float(t["amount"]) for t in expenses if not _is_exam_week_transaction(t)]
    delta = _mean(exam_expenses) - _mean(non_exam_expenses)
    exam_week_delta = delta if math.isfinite(delta) else 0.0

    return HabitProfile(
        avg_spend_by_day_of_week=avg_by_dow,
        avg_spend_by_month=avg_by_month,
        exam_week_delta=exam_week_delta,
    )


def detect_anomalies(
    transactions: list[dict[str, Any]], z_threshold: float = 2.0
) -> list[SpendingAnomaly]:
    """Return transactions that are statistical outliers within their category."""
    expenses = _only_expenses(transactions)
    stats = _get_category_stats(expenses)
    anomalies: list[SpendingAnomaly] = []

    for tx in expenses:
        cat = _normalize_category(tx.get("category"))
        s = stats.get(cat, {})
        std = s.get("std_dev", 0)
        if std == 0:
            continue

        z = (float(tx.get("amount", 0)) - s["mean"]) / std
        if abs(z) >= z_threshold:
            anomalies.append(
                SpendingAnomaly(
                    transaction_id=str(tx.get("id", "")),
                    category=cat,
                    amount=float(tx.get("amount", 0)),
                    z_score=z,
                )
            )

    return anomalies


def predict_habitual_expenses(
    transactions: list[dict[str, Any]],
    future_dates: list[datetime],
) -> list[HabitualPrediction]:
    """Predict daily spend for each future date based on learned habits."""
    habits = learn_habits(transactions)
    expenses = _only_expenses(transactions)

    now = datetime.now(tz=timezone.utc)
    cutoff = now - timedelta(days=30)
    recent_amounts = [
        float(t["amount"])
        for t in expenses
        if _to_datetime(t.get("date")) >= cutoff
    ]
    recent_avg = _mean(recent_amounts)

    predictions: list[HabitualPrediction] = []
    for dt in future_dates:
        # Python weekday(): Mon=0 … Sun=6; JS getDay(): Sun=0 … Sat=6
        # Map Python weekday to JS convention for consistency
        py_wd = dt.weekday()
        js_wd = (py_wd + 1) % 7

        day_mean = habits.avg_spend_by_day_of_week.get(js_wd, 0)
        month_mean = habits.avg_spend_by_month.get(dt.month - 1, 0)
        baseline = (day_mean + month_mean + recent_avg) / 3

        # Weekend exam-week boost (Sat/Sun in JS = 0 or 6)
        exam_boost = (
            habits.exam_week_delta * 0.15
            if habits.exam_week_delta > 0 and js_wd in (0, 6)
            else 0.0
        )

        predictions.append(
            HabitualPrediction(
                date=dt.isoformat(),
                predicted_spend=max(0.0, baseline + exam_boost),
            )
        )

    return predictions


def forecast_by_category(
    transactions: list[dict[str, Any]], days: int = 30
) -> list[CategoryForecast]:
    """Forecast spending per category for the next `days` days."""
    expenses = _only_expenses(transactions)
    now = datetime.now(tz=timezone.utc)
    window_start = now - timedelta(days=90)

    grouped: dict[str, list[float]] = {}
    for tx in expenses:
        dt = _to_datetime(tx.get("date"))
        if dt < window_start:
            continue
        cat = _normalize_category(tx.get("category"))
        grouped.setdefault(cat, []).append(float(tx.get("amount", 0)))

    day_multiplier = days / 30
    forecasts = [
        CategoryForecast(
            category=cat,
            amount=_mean(vals) * max(1, len(vals) / 6) * day_multiplier,
        )
        for cat, vals in grouped.items()
    ]
    forecasts.sort(key=lambda f: f.amount, reverse=True)
    return forecasts


def generate_future_dates(days: int = 30) -> list[datetime]:
    now = datetime.now(tz=timezone.utc)
    return [now + timedelta(days=i + 1) for i in range(days)]

File: src/lib/behavioralModel.ts
import { Transaction } from "@/types";

export interface HabitProfile {
  avgSpendByDayOfWeek: Record<number, number>;
  avgSpendByMonth: Record<number, number>;
  examWeekDelta: number;
}

export interface CategoryForecast {
  category: string;
  amount: number;
}

export interface HabitualPrediction {
  date: string;
  predictedSpend: number;
}

export interface SpendingAnomaly {
  transactionId: string;
  category: string;
  amount: number;
  zScore: number;
}

interface CategorizedSpendStats {
  mean: number;
  stdDev: number;
}

const EXAM_WEEK_KEYWORDS = ["exam", "midterm", "final", "quiz", "project"];

function toDate(value: string | Date): Date {
  return value instanceof Date ? value : new Date(value);
}

function mean(values: number[]): number {
  if (!values.length) return 0;
  return values.reduce((sum, v) => sum + v, 0) / values.length;
}

function stdDev(values: number[]): number {
  if (values.length < 2) return 0;
  const avg = mean(values);
  const variance = values.reduce((sum, v) => sum + (v - avg) ** 2, 0) / values.length;
  return Math.sqrt(variance);
}

function onlyExpenseTransactions(transactions: Transaction[]): Transaction[] {
  return transactions.filter((tx) => tx.type === "EXPENSE");
}

function normalizeCategory(category?: string): string {
  return (category || "Other").trim();
}

function getCategoryStats(transactions: Transaction[]): Record<string, CategorizedSpendStats> {
  const grouped: Record<string, number[]> = {};
  for (const tx of transactions) {
    const category = normalizeCategory(tx.category);
    if (!grouped[category]) grouped[category] = [];
    grouped[category].push(tx.amount);
  }

  const stats: Record<string, CategorizedSpendStats> = {};
  for (const [category, values] of Object.entries(grouped)) {
    stats[category] = {
      mean: mean(values),
      stdDev: stdDev(values),
    };
  }
  return stats;
}

function isExamWeekTransaction(tx: Transaction): boolean {
  const text = `${tx.description} ${tx.category || ""}`.toLowerCase();
  return EXAM_WEEK_KEYWORDS.some((keyword) => text.includes(keyword));
}

export function learnHabits(transactions: Transaction[]): HabitProfile {
  const expenses = onlyExpenseTransactions(transactions);
  const dayBuckets: Record<number, number[]> = {};
  const monthBuckets: Record<number, number[]> = {};

  for (const tx of expenses) {
    const date = toDate(tx.date);
    const day = date.getDay();
    const month = date.getMonth();

    if (!dayBuckets[day]) dayBuckets[day] = [];
    if (!monthBuckets[month]) monthBuckets[month] = [];

    dayBuckets[day].push(tx.amount);
    monthBuckets[month].push(tx.amount);
  }

  const avgSpendByDayOfWeek: Record<number, number> = {};
  const avgSpendByMonth: Record<number, number> = {};

  for (let i = 0; i < 7; i += 1) {
    avgSpendByDayOfWeek[i] = mean(dayBuckets[i] || []);
  }
  for (let i = 0; i < 12; i += 1) {
    avgSpendByMonth[i] = mean(monthBuckets[i] || []);
  }

  const examWeekExpenses = expenses.filter(isExamWeekTransaction).map((tx) => tx.amount);
  const nonExamExpenses = expenses.filter((tx) => !isExamWeekTransaction(tx)).map((tx) => tx.amount);
  const examWeekDelta = mean(examWeekExpenses) - mean(nonExamExpenses);

  return {
    avgSpendByDayOfWeek,
    avgSpendByMonth,
    examWeekDelta: Number.isFinite(examWeekDelta) ? examWeekDelta : 0,
  };
}

export function detectAnomalies(transactions: Transaction[], zThreshold = 2): SpendingAnomaly[] {
  const expenses = onlyExpenseTransactions(transactions);
  const statsByCategory = getCategoryStats(expenses);
  const anomalies: SpendingAnomaly[] = [];

  for (const tx of expenses) {
    const category = normalizeCategory(tx.category);
    const stats = statsByCategory[category];
    if (!stats || stats.stdDev === 0) continue;

    const zScore = (tx.amount - stats.mean) / stats.stdDev;
    if (Math.abs(zScore) >= zThreshold) {
      anomalies.push({
        transactionId: tx.id,
        category,
        amount: tx.amount,
        zScore,
      });
    }
  }

  return anomalies;
}

export function predictHabitualExpenses(
  transactions: Transaction[],
  futureDates: Date[]
): HabitualPrediction[] {
  const habits = learnHabits(transactions);
  const expenses = onlyExpenseTransactions(transactions);
  const recentAvg = mean(
    expenses
      .filter((tx) => {
        const date = toDate(tx.date);
        const now = new Date();
        const diffMs = now.getTime() - date.getTime();
        return diffMs <= 30 * 24 * 60 * 60 * 1000;
      })
      .map((tx) => tx.amount)
  );

  return futureDates.map((date) => {
    const dayMean = habits.avgSpendByDayOfWeek[date.getDay()] || 0;
    const monthMean = habits.avgSpendByMonth[date.getMonth()] || 0;
    const baseline = (dayMean + monthMean + recentAvg) / 3;
    const examBoost = habits.examWeekDelta > 0 && [0, 6].includes(date.getDay())
      ? habits.examWeekDelta * 0.15
      : 0;

    return {
      date: date.toISOString(),
      predictedSpend: Math.max(0, baseline + examBoost),
    };
  });
}

export function forecastByCategory(
  transactions: Transaction[],
  days = 30
): CategoryForecast[] {
  const expenses = onlyExpenseTransactions(transactions);
  const grouped: Record<string, number[]> = {};
  const now = new Date();
  const windowStart = new Date(now);
  windowStart.setDate(now.getDate() - 90);

  for (const tx of expenses) {
    const date = toDate(tx.date);
    if (date < windowStart) continue;
    const category = normalizeCategory(tx.category);
    if (!grouped[category]) grouped[category] = [];
    grouped[category].push(tx.amount);
  }

  const dayMultiplier = days / 30;
  return Object.entries(grouped)
    .map(([category, values]) => ({
      category,
      amount: mean(values) * Math.max(1, values.length / 6) * dayMultiplier,
    }))
    .sort((a, b) => b.amount - a.amount);
}

export function generateFutureDates(days = 30): Date[] {
  const dates: Date[] = [];
  const now = new Date();
  for (let i = 1; i <= days; i += 1) {
    const date = new Date(now);
    date.setDate(now.getDate() + i);
    dates.push(date);
  }
  return dates;
}
